osu.Game/Screens/Play/HUD/ArgonCounterTextComponent.cs

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Localisation;
+using osu.Framework.Text;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Skinning;
+using osuTK;
+
+namespace osu.Game.Screens.Play.HUD
+{
+    public partial class ArgonCounterTextComponent : CompositeDrawable, IHasText
+    {
+        private readonly ArgonCounterSpriteText wireframesPart;
+        private readonly ArgonCounterSpriteText textPart;
+        private readonly OsuSpriteText labelText;
+
+        public IBindable<float> WireframeOpacity { get; } = new BindableFloat();
+        public Bindable<int> RequiredDisplayDigits { get; } = new BindableInt();
+
+        public Container NumberContainer { get; private set; }
+
+        public LocalisableString Text
+        {
+            get => textPart.Text;
+            set
+            {
+                int remainingCount = RequiredDisplayDigits.Value - value.ToString().Count(char.IsDigit);
+                string remainingText = remainingCount > 0 ? new string('#', remainingCount) : string.Empty;
```

Comment:
We might need to reconsider these implementation from a performance perspective in the next profiling round. These string functions can be quite expensive when this text is being updated every update frame.

Reply:
I've spent some time benchmarking this with various methods of allocating a string, and ended up with the following results:
| Method                         | Count | Mean         | Error       | StdDev      | Gen0     | Gen1    | Allocated |
|------------------------------- |------ |-------------:|------------:|------------:|---------:|--------:|----------:|
| AllocateWithStringConstructor  | 1     |     198.7 ns |     2.44 ns |     2.28 ns |   0.2487 |       - |     520 B |
| AllocateWithStringBuilder      | 1     |     200.3 ns |     3.64 ns |     3.04 ns |   0.2487 |       - |     520 B |
| AllocateWithStringCreateMethod | 1     |     290.3 ns |     2.06 ns |     1.72 ns |   0.1988 |       - |     416 B |
| AllocateAndCache | 1     |    378.5 ns |   7.45 ns |   8.28 ns |  0.6733 |   1.38 KB |
| AllocateWithStringConstructor  | 10    |   1,667.8 ns |    33.02 ns |    46.30 ns |   1.9341 |       - |    4048 B |
| AllocateWithStringBuilder      | 10    |   1,629.6 ns |    16.52 ns |    14.64 ns |   1.9341 |       - |    4048 B |
| AllocateWithStringCreateMethod | 10    |   2,790.3 ns |    25.84 ns |    24.17 ns |   1.8845 |       - |    3944 B |
| AllocateAndCache | 10    |    920.1 ns |  17.87 ns |  16.71 ns |  1.0166 |   2.08 KB |
| AllocateWithStringConstructor  | 100   |  16,720.6 ns |   320.20 ns |   393.23 ns |  18.7988 |       - |   39328 B |
| AllocateWithStringBuilder      | 100   |  16,475.0 ns |   249.08 ns |   286.85 ns |  18.7988 |       - |   39328 B |
| AllocateWithStringCreateMethod | 100   |  29,170.3 ns |   362.28 ns |   302.52 ns |  18.7378 |       - |   39224 B |
| AllocateAndCache | 100   |  6,215.7 ns | 124.16 ns | 174.06 ns |  4.4479 |   9.11 KB |
| AllocateWithStringConstructor  | 1000  | 175,346.1 ns | 2,218.49 ns | 2,075.18 ns | 133.5449 | 35.1563 |  392128 B |
| AllocateWithStringBuilder      | 1000  | 172,832.6 ns | 1,270.34 ns |   991.80 ns | 133.0566 | 34.9121 |  392128 B |
| AllocateWithStringCreateMethod | 1000  | 291,833.8 ns | 5,300.95 ns | 4,958.51 ns |  82.0313 | 33.6914 |  392024 B |
| AllocateAndCache | 1000  | 60,663.7 ns | 875.24 ns | 730.86 ns | 38.4521 |  79.42 KB |

<details>
<summary>code</summary>

```csharp
// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
// See the LICENCE file in the repository root for full licence text.

using System.Collections.Generic;
using System.Text;
using BenchmarkDotNet.Attributes;

namespace osu.Game.Benchmarks
{
    public class BenchmarkStrings : BenchmarkTest
    {
        [Params(1, 10, 100, 1000)]
        public int Count { get; set; }

        [Benchmark]
        public string[] AllocateWithStringConstructor()
        {
            StringBuilder builder = new StringBuilder();
            string[] array = new string[Count * 10];

            for (int i = 0; i < Count; i++)
            {
                for (int j = 0; j < 10; j++)
                {
                    builder.Clear();
                    builder.Append('#', j);
                    array[i * 10 + j] = builder.ToString();
                }
            }

            return array;
        }

        [Benchmark]
        public string[] AllocateWithStringBuilder()
        {
            StringBuilder builder = new StringBuilder();
            string[] array = new string[Count * 10];

            for (int i = 0; i < Count; i++)
            {
                for (int j = 0; j < 10; j++)
                {
                    builder.Clear();
                    builder.Append('#', j);
                    array[i * 10 + j] = builder.ToString();
                }
            }

            return array;
        }

        [Benchmark]
        public string[] AllocateWithStringCreateMethod()
        {
            string[] array = new string[Count * 10];

            for (int i = 0; i < Count; i++)
            {
                for (int j = 0; j < 10; j++)
                    array[i * 10 + j] = string.Create(j, '#', (span, state) => span.Fill(state));
            }

            return array;
        }

        [Benchmark]
        public string[] AllocateAndCache()
        {
            Dictionary<int, string> dictionary = new Dictionary<int, string>();
            string[] array = new string[Count * 10];

            for (int i = 0; i < Count; i++)
            {
                for (int j = 0; j < 10; j++)
                {
                    if (!dictionary.TryGetValue(j, out string? value))
                        dictionary[j] = value = new string('#', j);

                    array[i * 10 + j] = value;
                }
            }

            return array;
        }
    }
}
```
</details>

With allocating strings of length 0-10 constant characters over thousand iterations, time sits around 0.175ms, so as far as I understand, it takes around ~0.000175ms on average to allocate a single string per frame, which should be safe enough to not have to add some sort of caching for, but you're more knowledgable in this area than I do.

Reply:
it's not cpu but memory concerns (which leads to GC etc.) so look at the "allocated" column if anything.